// classeviva.js/dist/classes/Rest.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const fs_1 = require("fs");
const Enums = __importStar(require("../base/enums"));
const client_1 = __importDefault(require("../base/client"));
class Rest extends client_1.default {
    username;
    #password;
    #token;
    #state;
    #directory;
    login_timeout;
    expiration;
    debug;
    saveTempFile;
    keepAlive;
    user;
    constructor(opts = {}) {
        super({
            debug: opts.debug || false,
            app: opts.app || Enums.Apps.Students,
            host: 'https://web.spaggiari.eu/'
        });
        this.username = opts.username || "";
        this.#password = opts.password || "";
        this.#state = opts.state || Enums.States.Italy;
        this.#directory = (0, path_1.join)((0, path_1.parse)(__dirname).dir, '..');
        this.login_timeout = null;
        this.debug = opts.debug || false;
        this.saveTempFile = opts.saveTempFile ?? true;
        this.keepAlive = opts.keepAlive || true;
        this.resetAuth();
        this.setHeaders({
            "Z-Dev-Apikey": "Tg1NWEwNGIgIC0K",
            "Z-If-None-Match": "",
        });
        this.log('app', this.app);
    }
    /**
     * Logins to Classeviva
     * @param {string} [username] Classeviva credentials username
     * @param {string} [password] Classeviva credentials password
     * @returns {object} user object
     */
    async login(username = this.username, password = this.#password) {
        if (this.authorized)
            return this.error("Already logged in");
        if (!username || !password)
            return this.error("Username or password not set");
        if (!await this.#checkTemp() || !this.saveTempFile) {
            const userData = {
                uid: username,
                pass: password,
            };
            const response = await this.httpClient.request({
                path: `${this.getPath()}/auth/login/`,
                method: "POST",
                headers: this.headers,
                body: JSON.stringify(userData),
            });
            const json = await response.body.json();
            if ("error" in json) {
                this.authorized = false;
                return this.error(json.message ?? json.error, json.statusCode);
            }
            if (response.statusCode !== 200)
                return this.error(`The server returned a status code other than 200`, response.statusCode);
            this.#updateData(json);
            if (this.saveTempFile) {
                await (0, fs_1.writeFileSync)(`${this.#directory}/cvv.json`, JSON.stringify(json, null, 2));
                this.log("Saved temp file ✅");
            }
        }
        if (!this.authorized)
            return this.error("Failed to login");
        this.log(`Successfully logged in as "${this.user.name} ${this.user.surname}" ✅`);
        if (this.keepAlive) {
            this.login_timeout = setTimeout(() => {
                this.login();
            }, 1000 * 60 * 60 * 1.5);
            this.log(`Set login_timeout`);
        }
        ;
        return this.user;
    }
    /**
     * Logs out from Classeviva
     * @returns {boolean} true if logged out, error if already logged out
     */
    logout() {
        if (!this.authorized)
            return true;
        if (this.login_timeout)
            clearTimeout(this.login_timeout);
        this.resetAuth();
        this.log("Successfully logged out ✅");
        return true;
    }
    /**
     * Get student's cards
     * @returns {object[]} Array of objects containing the student's cards
     */
    async getCards() {
        const data = await this.#fetchRest({ path: "/cards" });
        if (data?.cards && data?.cards?.length > 0)
            this.#updateUser(data.cards[0]);
        return data?.cards ?? [];
    }
    /**
     * Get student's card
     * @returns {object} Object containing the student's card
     */
    async getCard() {
        const data = await this.#fetchRest({ path: "/card" });
        if (data?.card && Object.keys(data?.card ?? {}).length > 0)
            this.#updateUser(data.card);
        return data?.card;
    }
    /**
     * Get student's grades
     * @returns {object[]} Array of objects containing the student's grades
     */
    async getGrades() {
        const data = await this.#fetchRest({ path: `/grades2` });
        return data?.grades ?? [];
    }
    /**
     * Get student's absences
     * @returns {object[]} Array of objects containing the student's absences
     */
    async getAbsences() {
        const data = await this.#fetchRest({ path: `/absences/details` });
        return data?.events ?? [];
    }
    /**
     * Get student's agenda
     * @param {string} filter "all" | "homework" | "other", default "all", used to filter the agenda
     * @param {Date} start The start date of the agenda (defaults to today)
     * @param {Date} end  The end date of the agenda (defaults to today)
     * @returns {object[]} Array of objects containing the student's agenda
     */
    async getAgenda(filter = "all", start = new Date(), end = new Date()) {
        const filters = ["all", "homework", "other"];
        if (!filters.includes(filter))
            return this.error("Invalid filter");
        const map = {
            all: "all",
            homework: "AGHW",
            other: "AGNT",
        };
        const data = await this.#fetchRest({ path: `/agenda/${map[filter]}/${this.#formatDate(start)}/${this.#formatDate(end)}` });
        return data?.agenda ?? [];
    }
    /**
     * Get student's documents
     * @returns {object[]} Array of objects containing the student's documents
     */
    async getDocuments() {
        const data = await this.#fetchRest({ path: "/documents", method: "POST" });
        return data ?? [];
    }
    /**
     * Get student's noticeboard items
     * @returns {object[]} Array of objects containing the student's noticeboard items
     */
    async getNoticeboard() {
        const data = await this.#fetchRest({ path: "/noticeboard" });
        return data?.items ?? [];
    }
    /**
     * Get student's books
     * @returns {object[]} Array of objects containing the student's books
     */
    async getSchoolBooks() {
        const data = await this.#fetchRest({ path: "/schoolbooks" });
        return data?.schoolbooks;
    }
    /**
     * Get student's calendar
     * @returns {object[]} Array of objects containing the student's calendar
     */
    async getCalendar() {
        const data = await this.#fetchRest({ path: "/calendar/all" });
        return data?.calendar ?? [];
    }
    /**
     * Get student's lessons
     * @param {boolean} [today] Boolean to get today's lessons, default true
     * @param {Date} [start] If today is false, the start date of the lessons (defaults to today)
     * @param {Date} [end] If today is false, the end date of the lessons (defaults to today)
     * @returns {object[]} Array of objects containing the student's lessons
     */
    async getLessons(today = true, start = new Date(), end = new Date()) {
        const data = await this.#fetchRest({ path: `/lessons${today ? "/today" : `/${this.#formatDate(start)}/${this.#formatDate(end)}`}` });
        return data?.lessons ?? [];
    }
    /**
     * Get student's notes
     * @returns {object[]} Array of objects containing the student's notes
     */
    async getNotes() {
        const data = await this.#fetchRest({ path: "/notes/all" });
        return data ?? [];
    }
    /**
     * Get student's periods
     * @returns {object[]} Array of objects containing the student's periods
     */
    async getPeriods() {
        const data = await this.#fetchRest({ path: "/periods" });
        return data?.periods ?? [];
    }
    /**
     * Get student's subjects
     * @returns {object[]} Array of objects containing the student's subjects
     */
    async getSubjects() {
        const data = await this.#fetchRest({ path: "/subjects" });
        return data?.subjects ?? [];
    }
    /**
     * Get student's didactics items
     * @returns {object[]} Array of objects containing the student's didactics items
     */
    async getDidactics() {
        const data = await this.#fetchRest({ path: "/didactics" });
        return data?.didacticts ?? [];
    }
    /**
     * Get a list of the possible parents options for classeviva
     * @returns {object} An object containing all the possible parents options for classeviva
     */
    async getParentsOptions() {
        const data = await this.#fetchRest({ path: "/_options", method: "GET", type: "parents" });
        return data?.options ?? {};
    }
    /**
     *  Get a list of the avaible talks with teachers on classeviva
     * @returns {object[]} An array of objects containing data about the avaible talks with teachers for classeviva
     */
    async getOverallTalks() {
        const data = await this.#fetchRest({ path: "/overalltalks/list", method: "GET", type: "parents" });
        return data?.overallTalks ?? [];
    }
    /**
     *  Get a list of something regarding the talks with teachers
     * @param {Date} start The start date of the talks (defaults to today)
     * @param {Date} end The end date of the talks (defaults to today)
     * @returns {object[]} An array of objects containing data about the talks with teachers for classeviva
     */
    async getTalks(start = new Date(), end = new Date()) {
        const data = await this.#fetchRest({ path: `/talks/teachersframes/${this.#formatDate(start)}/${this.#formatDate(end)}`, method: "GET", type: "parents" });
        return data?.teachers ?? [];
    }
    /**
     *  Get auth ticket
     * @returns {object} An object containing data about the auth ticket
     */
    async getTicket() {
        return this.fetch({
            url: `${this.getPath()}/auth/ticket`
        });
    }
    /**
     *  Get the user avatar
     * @returns {unknown} The user avatar (not tested)
     */
    async getAvatar() {
        return this.fetch({
            url: `${this.getPath()}/auth/avatar`
        });
    }
    /**
     * Get an overview of the day specified or the time specified
     * @param {Date} start The start date of the overview (defaults to today)
     * @param {Date} end The end date of the overview (defaults to today)
     * @returns {object} An object containing data about the overview of a day or the time specified
     */
    async getOverview(start = new Date(), end = new Date()) {
        const data = await this.#fetchRest({ path: `/overview/all/${this.#formatDate(start)}/${this.#formatDate(end)}` });
        return data;
    }
    /*/**
     * Not implemented yet 😢
     * @param {string || number} bookingId booking id of the talk
     * @param {string} message message to send
     * @returns {unknown}
     */
    /*async sendTeacherMessage(bookingId: string, message: string) {
        const data = await this.#fetchRest(`/talks/teachermessage/${bookingId}`, "POST", "parents");
        return data ?? {};
    }; */
    /**
     * Read messages from the inbox of a talk
     * @param {string || number} bookingId booking id of the talk
     * @param {string} message message to send
     * @returns {object}
     */
    async readTalkMessage(bookingId) {
        const data = await this.#fetchRest({ path: `/talks/teachermessage/${bookingId}`, method: "POST", type: "parents", body: JSON.stringify({ "messageRead": true }) });
        return data ?? {};
    }
    /**
     * Checks if a document is avaible
     * @param {string | number} hash The hash of the document
     * @returns {object} An object containing data about the document
     */
    async checkDocument(hash) {
        const data = await this.#fetchRest({ path: `/documents/check/${hash}/`, method: "POST" });
        return data;
    }
    /**
     * Book a talk with a teacher
     * @param {string | number} teacherId The id of the teacher
     * @param {string | number} talkId The id of the talk
     * @param {string | number} slot The slot of the talk
     * @param {object} opts contact options
     * @returns {object} An object containing data about the booked talk
     */
    async bookTalk(teacherId, talkId, slot, opts) {
        const data = await this.#fetchRest({ path: `/talks/book/${teacherId}/${talkId}/${slot}`, method: "POST", type: "parents", body: JSON.stringify(opts) });
        return data ?? {};
    }
    /**
     * Get the list of contents that's displayed in the app (should be "Classeviva extra")
     * @param {number} year year of contents
     * @param {boolean} common idk, defaults to true
     * @returns {object[]} An array of objects containing data about the contents that's displayed in the app
     */
    async getContents(year = new Date().getFullYear(), common = true) {
        if (!this.user.school?.code)
            return this.error("No school code, please update using getCard() or getCards()");
        return this.fetch({
            url: `${this.getHost()}/gek/api/v1/${this.user.school.code}/${year}/students/contents?common=${common}`
        });
    }
    /**
     * Get infos about your agreement to the terms of classeviva. If you haven't agreed yet, this response body will be empty and the function will return an empty object.
     * @returns {object} An object containing data about the agreement to the terms of classeviva
     */
    async getTermsAgreement() {
        const data = await this.#fetchRest({ path: "/getTermsAgreement", type: "users", id: "userIdent" });
        return data;
    }
    /**
     * Set the agreement to the terms of classeviva third party data colletors
     * @param {boolean} ThirdParty Whether you agree to the terms of classeviva third party data colletors, defaults to true
     * @returns {object} An object with the property "msg": "ok" if the agreement was set successfully
     */
    async setTermsAgreement(ThirdParty = false) {
        const accepted = ThirdParty ? "1" : "0";
        const data = await this.#fetchRest({ path: "/setTermsAgreement", method: "POST", type: "users", body: JSON.stringify({ bitmask: accepted }), id: "userIdent" });
        return data ?? { msg: "NOT OK" };
    }
    /**
     * Read a notice from the school
     * @param {string} eventCode Event code of the notice
     * @param {string | number} id Id of the notice
     * @param {object} options { sign, join, text }
     * @returns {object} An object containing data about the notice
     */
    async readNotice(eventCode, id, options = {}) {
        const data = await this.#fetchRest({
            path: `/noticeboard/read/${eventCode}/${id}/101`, method: "POST", body: options ? JSON.stringify(options) : "", customHeaders: {
                "Content-Type": "application/x-www-form-urlencoded"
            }
        });
        return data;
    }
    /**
     * Get the document url of a notice attachment
     * @param {string} eventCode Event code of the notice
     * @param {string | number} id Id of the notice
     * @returns {string} The url of the document
     */
    async getNoticeDocumentUrl(eventCode, id) {
        if (!this.authorized)
            return this.error("Not authorized");
        const headers = {
            ...this.headers,
            "Z-Auth-Token": this.#token,
        };
        const response = await this.httpClient.request({
            path: `${this.getPath()}/students/${this.user.ident}/noticeboard/attach/${eventCode}/${id}/`,
            method: 'GET',
            headers,
        });
        const location = response.headers["Location"];
        const url = Array.isArray(location) ? location[0] : location;
        return url ?? "";
    }
    /**
     * Get the status of the user token
     * @returns {object} An object containing data about the token
     */
    async getTokenStatus() {
        return this.fetch({
            url: `${this.getPath()}/auth/status`
        });
    }
    /**
     * Read a document
     * @param {string} hash the hash of the document
     * @returns {Buffer} The document
     */
    async readDocument(hash) {
        const data = await this.#fetchRest({ path: `/documents/read/${hash}/`, method: "POST", responseType: "buffer" });
        return data;
    }
    async resetPassword(email) {
        return this.fetch({
            url: `${this.getHost()}sso/app/default/sam.php?a=akRSPWRQ`,
            body: `eml=${email}`,
            customHeaders: {
                "Content-Type": "application/x-www-form-urlencoded",
                "X-Requested-With": "XMLHttpRequest"
            }
        });
    }
    async readNote(noteType, noteId) {
        const data = await this.#fetchRest({ path: `/notes/${noteType}/read/${noteId}/`, method: "POST", type: "students", body: undefined });
        return data?.event;
    }
    async getTerms() {
        return this.fetch({
            url: `${this.getHost()}auc/api/v2/getTerms`,
        });
    }
    async getAucContents() {
        return this.fetch({
            url: `${this.getHost()}auc/api/v2/contents`,
            responseType: "text",
        });
    }
    async getAucContentAuths() {
        return this.fetch({
            url: `${this.getHost()}auc/api/v2/contentAuths`,
            responseType: "text",
        });
    }
    async getSchoolPresentation(schoolCode) {
        return this.fetch({
            url: `${this.getHost()}gek/getSchoolPresentation/${schoolCode}`,
            responseType: "text",
        });
    }
    setState(newState) {
        this.#state = newState;
        this.rebuildHTTPClient();
        this.log('set state to', newState);
    }
    ;
    async getHomeworks() {
        return this.#fetchRest({ path: '/homeworks' });
    }
    async getMinigameToken() {
        /*
            minigameToken payload:
            {
                "exp": number, //expiration unix timestamp
                "sub": string, //user identi (with letters)
                "aud": string, //school code
                "syr": string //year
            }
        */
        return this.fetch({
            url: `${this.getPath()}/auth/minigame`
        });
    }
    async getMinigameLeaderboard(scope, gameId) {
        return await this.#fetchMinigame({
            url: `${this.getPath()}/minigame/leaderboards/${scope}/${gameId}/${this.user.ident}`,
        });
    }
    async checkSchool(schoolCode = this.user.school?.code, year = 2022) {
        return await this.fetch({
            url: `${this.getHost()}gek/api/v1/${schoolCode}/${year}/checkSchool`,
        });
    }
    getPath() {
        return `${this.getHost()}rest/v1`;
    }
    /**
     * @private Returns the classeviva host
     * @returns {string} The classeviva host
     */
    getHost(state = this.#state) {
        return `https://${Enums.StateUrls[state]}/`;
    }
    /**
     * @private Updates the user object with school infos and the user type
     * @param {object} card The user card (from getCard() or getCards())
     * @return {void} Nothing
     */
    #updateUser(card) {
        const { schName, schDedication, schCity, schProv, schCode, usrType } = card;
        this.user.type = Enums.UserTypes[usrType || "S"];
        this.user.school = {
            name: schName,
            dedication: schDedication,
            city: schCity,
            province: schProv,
            code: schCode
        };
    }
    /**
     * @private Checks for temp file
     * @returns {boolean} True if there was temp file and it updated the data, false otherwise or in case of an error
     */
    async #checkTemp() {
        try {
            const temp = await (0, fs_1.readFileSync)(`${this.#directory}/cvv.json`, "utf8");
            if (!temp) {
                return false;
            }
            const json = JSON.parse(temp);
            if (new Date(json.expire) <= new Date())
                return false;
            this.#updateData(json);
            return true;
        }
        catch (e) {
            return false;
        }
    }
    /**
     * @private Updates user data
     * @param {LoginResponse} data user data to update
     * @returns {void}
     */
    #updateData(data) {
        if (!data) {
            this.logout();
            return;
        }
        if (typeof data !== "object")
            return;
        new Date(data.expire || new Date()) > new Date() ? this.authorized = true : this.authorized = false;
        this.#token = data.token || "";
        this.user = {
            name: data.firstName,
            surname: data.lastName,
            id: this.#removeLetters(data.ident || ""),
            ident: data.ident,
        };
        this.expiration = data.expire || `${new Date().getFullYear()}-${new Date().getMonth()}-${new Date().getDate()}T${new Date().getHours()}:${new Date().getMinutes()}:${new Date().getSeconds()}+01:00`;
        return;
    }
    /**
     * @private Removes letters from a string
     * @param {string} string string to remove letters from
     * @returns {string} string without letters
     */
    #removeLetters(string) {
        return string.replace(/[^0-9]/g, "");
    }
    /**
     * @private Formats date to string in format YYYYMMDD
     * @param {Date} date date to format
     * @returns {string} date in format YYYYMMDD
     */
    #formatDate(date = new Date()) {
        const year = date.getFullYear();
        const month = date.getMonth() + 1;
        const day = date.getDate();
        return `${year}${month < 10 ? "0" + month : month}${day < 10 ? "0" + day : day}`;
    }
    async fetch({ url, method = "GET", body, responseType = "json", customHeaders = {} }) {
        if (!this.authorized)
            return this.error("Not logged in");
        const headers = {
            ...this.headers,
            "Z-Auth-Token": this.#token,
            ...customHeaders,
        };
        const options = {
            path: url,
            method,
            headers,
            body: method.toUpperCase() !== 'GET' ? body : undefined
        };
        const response = await this.httpClient.request(options);
        switch (responseType) {
            case 'json':
                const data = await response.body.json();
                if ("error" in data) {
                    return this.error(data.message ? data.message : (data.error.split('/').pop() ?? 'Unknown error'), data.statusCode);
                }
                return data;
            case 'buffer':
                return await response.body.blob();
            case 'text':
                return await response.body.text();
            default:
                return this.error('Invalid responseType');
        }
    }
    /**
     * @private Fetch data from the REST API
     * @param {string} path api path
     * @param {string} [method] http method
     * @param {string} [type] students | parents
     * @param {string} [body] body to send
     * @param {string} [responseType] response body type
     * @param {string} [id] user identifier
     * @param {object} [customHeaders] additional headers to send
     * @returns {Promise<unknown>} the response
     */
    async #fetchRest({ path = "/", type = "students", id = "userId", ...opts }) {
        const url = `${this.getPath()}/${type}/${id == "userId" ? this.user.id : this.user.ident}${path}`;
        return this.fetch({
            url,
            ...opts,
        });
    }
    async #fetchMinigame(opts) {
        const data = await this.getMinigameToken();
        if (!data?.minigameToken)
            return;
        return this.fetch({
            ...opts,
            customHeaders: {
                'Authorization': `Bearer ${data.minigameToken}`,
            }
        });
    }
    resetAuth() {
        this.authorized = false;
        this.#token = "";
        this.expiration = "";
        this.user = {
            name: undefined,
            surname: undefined,
            id: undefined,
            ident: undefined,
            type: undefined,
            school: {}
        };
        return this;
    }
    ;
}
exports.default = Rest;
