// classeviva.js/dist/base/client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const http_1 = __importDefault(require("./http"));
const api_1 = __importDefault(require("../errors/api"));
class BaseApiClient {
    headers;
    httpClient;
    debug;
    app;
    authorized;
    constructor(options) {
        this.debug = options.debug;
        if (options.log)
            this.log = options.log;
        if (options.resetAuth)
            this.resetAuth = options.resetAuth;
        this.httpClient = new http_1.default({
            url: options.host ?? this.getHost(),
            debug: options.debug,
        });
        this.app = options.app || null;
        this.headers = {
            "Content-Type": "application/json",
            "Accept": "application/json",
            "User-Agent": this.app || '',
            "X-Requested-With": "XMLHttpRequest",
        };
    }
    log(...args) {
        if (!this.debug)
            return;
        console.log(`[\x1b[31mCLASSEVIVA\x1b[0m]`, ...args);
    }
    error(message, statusCode = 0) {
        this.log(`An error happened: ${message} ❌`);
        return Promise.reject(new api_1.default(message, statusCode));
    }
    getHost() {
        return 'https://web.spaggiari.eu/';
    }
    ;
    rebuildHTTPClient() {
        this.httpClient.close();
        this.httpClient = new http_1.default({
            url: this.getHost(),
            debug: this.debug,
        });
        return this;
    }
    setHeaders(headers) {
        Object.assign(this.headers, headers);
        return this;
    }
    getCookie(response, name) {
        const cookies = response.headers[name];
        if (!cookies)
            return "";
        const cookie = (Array.isArray(cookies) ? cookies[0] : cookies)
            .split(", ")
            .pop();
        return cookie ?? "";
    }
    /**
     * Get a list of the Classeviva class' functions
     * @returns {string[]} An array containing the Classeviva class' functions
     */
    getMethods() {
        return Object
            .getOwnPropertyNames(Object.getPrototypeOf(this))
            .filter(prop => prop !== "constructor");
    }
}
;
exports.default = BaseApiClient;
