// classeviva.js/dist/base/docs.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const fs_1 = __importDefault(require("fs"));
const docs_1 = __importDefault(require("../errors/docs"));
const tab = '   ';
class DocsUpdater {
    readmeString = "";
    async updateDocs() {
        this.log('📄', 'Updating docs...');
        const dir = (0, path_1.join)(process.cwd(), `docs`);
        if (!await this.existsOrCreate({ path: dir }))
            this.error('Missing docs dir');
        await this.classes(dir);
        await this.readme(dir, this.readmeString);
        await this.examples(dir);
        this.log('📃', 'Updated all files');
        this.readmeString = '';
        return this;
    }
    async classes(docsDir) {
        var _a;
        const distPath = (0, path_1.join)(process.cwd(), 'dist');
        if (!await this.existsOrCreate({ path: distPath, create: false }))
            return this.error('Missing dist folder, please compile');
        const classesPath = (0, path_1.join)(distPath, 'classes');
        if (!await this.existsOrCreate({ path: classesPath }))
            return this.error('Missing classes dir');
        const classes = fs_1.default
            .readdirSync(classesPath)
            .filter(file => file.endsWith('.js'));
        for (const cls of classes) {
            this.log(`🧱`, `Updating class "${cls}"...`);
            const clsPath = (0, path_1.join)(classesPath, cls);
            const clsModule = await (_a = clsPath, Promise.resolve().then(() => __importStar(require(_a))));
            const clsInstance = new clsModule.default();
            const data = this.getData(clsInstance);
            const files = Object.keys(data);
            const clsName = this.removeExtension(cls);
            const docClass = (0, path_1.join)(docsDir, clsName);
            await this.existsOrCreate({
                path: docClass,
                dir: true
            });
            this.readmeString += `\n• ${clsName}\n\n`;
            for (const file of files) {
                this.log('-', `Updating file "${file}"...`, true);
                const filePath = (0, path_1.join)(docClass, `${file}.md`);
                await this.existsOrCreate({
                    path: filePath,
                    dir: false
                });
                const content = fs_1.default.readFileSync(filePath, 'utf8');
                const split = content.split(`# ${file}`);
                const toReplace = split.length > 0 ? (split.pop() ?? '') : '';
                const newContent = data[file].join('\n');
                if (!toReplace.length) {
                    this.log('⚠️', `"${file}" is empty`, true);
                    fs_1.default.writeFileSync(filePath, `# ${file}\n\n${newContent}`);
                }
                else
                    fs_1.default.writeFileSync(filePath, content.replace(toReplace, `\n\n${newContent}`));
                this.readmeString += `${tab}- [${file}](${clsName}/${file}.md)\n`;
                this.log('🆗', `Updated "${file}"`, true);
            }
            this.log('✅', `Updated class "${cls}"`);
        }
        this.log('🌐', 'Updated all classes');
    }
    async readme(docsDir, newContent) {
        const path = (0, path_1.join)(docsDir, 'README.md');
        await this.existsOrCreate({
            path,
            dir: false,
        });
        let content = await fs_1.default.readFileSync(path, 'utf-8');
        if (!content)
            content = '# Docs\nHere you can find infos about the classes contained in this package. Choose a document from the list below:\n\n## Documents';
        const split = content.split(`## Documents`);
        const toReplace = split.length > 0 ? (split.pop() ?? '') : '';
        if (!toReplace.length) {
            this.log('⚠️', '"README.md" is empty');
            fs_1.default.writeFileSync(path, `${content}\n${newContent}`);
        }
        else {
            this.log('-', 'Updating "README.md"');
            fs_1.default.writeFileSync(path, content.replace(toReplace, `\n${newContent}`));
        }
        this.log('✅', 'Updated "README.md"', true);
    }
    async examples(docsDir) {
        this.log('📦', 'Updating examples...');
        const path = (0, path_1.join)(docsDir, 'Examples');
        await this.existsOrCreate({
            path,
            dir: true
        });
        let string = "";
        const examples = fs_1.default.readdirSync(path);
        examples.forEach(async (example) => {
            string += `\n• [${this.removeExtension(example)}](docs/Examples/${example})\n`;
            this.log(`🆗`, `Added example "${example}"`, true);
        });
        if (!string.length) {
            string = '\nThere are no examples yet.\n';
            this.log('⚠️', '"Examples" folder is empty', true);
        }
        const readmePath = (0, path_1.join)(process.cwd(), 'README.md');
        const readmeContent = await fs_1.default.readFileSync(readmePath, 'utf-8');
        const split = readmeContent.split(`## Examples`);
        if (split.length === 0)
            return this.error('Couldn\'t find "Examples" section in README.md');
        const toReplace = split.pop()?.split('##')[0] ?? '';
        fs_1.default.writeFileSync(readmePath, readmeContent.replace(toReplace, `\n${string}\n`));
        this.log('✅', 'Updated examples');
        return;
    }
    async existsOrCreate({ create = true, dir = true, ...options }) {
        const exists = await fs_1.default.existsSync(options.path);
        if (exists)
            return true;
        if (!create)
            return false;
        try {
            dir ? await fs_1.default.mkdirSync(options.path, {
                recursive: true
            }) : await fs_1.default.writeFileSync(options.path, '');
            return true;
        }
        catch (e) {
            this.error(`Couldn't create "${options.path}" path\n`);
        }
    }
    removeExtension(str) {
        return str.split('.')[0];
    }
    error(msg) {
        this.log('❌', msg);
        throw new docs_1.default(msg);
    }
    log(emoji, message, space = false) {
        console.log((space ? tab : '') + `[${emoji}]`, message);
        return this;
    }
    getData(cls) {
        return {
            Methods: cls
                .getMethods()
                .map((m) => `> ${m}\n`),
            Properties: Object
                .getOwnPropertyNames(cls)
                .map((prop) => `> ${prop}\n`),
        };
    }
}
;
exports.default = DocsUpdater;
